Point groups

patrykgruszka opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ N ]
| New feature/enhancement? (use the a.x branch)      | [ Y ]
| Deprecations?                          | [ N ]
| BC breaks? (use the c.x branch)        | [ N ]
| Automated tests included?              | [ Y ] <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | https://github.com/mautic/user-documentation/pull/152
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

#### Description:
This PR adds possibility to create point groups to categorize contact scores.

New features in this PR:
- Manage Groups (create, read, update, delete)
- Group usage in Point action
- Group usage in Point triggers
- View Group points in Contact details
- Campaign condition based on Group points
- Campaign action to modify Group points
- Segment filter based on Group points
- Group points Report

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:

<!--
This part is really important. If you want your PR to be merged, take the time to write very clear, annotated and step by step test instructions. Do not assume any previous knowledge - testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Load fixtures `ddev exec bin/console doctrine:fixtures:load`
3. Update segments `ddev exec bin/console m:s:u`
4. Add some point leagues (Points -> Manage Groups) Example:
![image](https://user-images.githubusercontent.com/8580942/227149236-662024d9-6087-4620-bdca-8df7080f4b1f.png)
5. Create new point action with Point Group, example:
![image](https://user-images.githubusercontent.com/8580942/227149765-be38d6b0-0583-4f3e-a6f3-11d1cd5162b5.png)
6. Create Point trigger with Group. In my example i will use add tag event to verify if it works fine
![image](https://user-images.githubusercontent.com/8580942/227152897-65e7b272-b546-46ee-b710-fc3473ee1671.png)
7. Create a segment email (remember to set Published at date and set as Published)
![image](https://user-images.githubusercontent.com/8580942/227151415-4021147d-d8f7-49ef-8a64-33fef17a23fa.png)
8. Trigger email broadcast send with command `ddev exec bin/console m:b:s`
![image](https://user-images.githubusercontent.com/8580942/227151593-2ff949ae-be61-4b24-b59b-a894b6b60b81.png)
9. Check that the contact who received the email has a Group score.
![image](https://user-images.githubusercontent.com/8580942/227152084-9a290cb3-4487-41a7-8a58-200883c1ac86.png)
10. Check that the contact who meet the trigger point requirements has tag added
11. Add new segment with Group filter
![image](https://user-images.githubusercontent.com/8580942/227154825-5b06266f-ba1c-40b9-ad93-87ce1cfd6278.png)
12. Update segments: `ddev exec bin/console m:s:u`
13. Check if contacts that meet the requirements were added to the segment
![image](https://user-images.githubusercontent.com/8580942/227155165-340b2e55-6d0d-4136-9a59-1ed0a0a9268e.png)
14. Create a campaign that starts from this segment
15. Add condition based on contact score with Group
![image](https://user-images.githubusercontent.com/8580942/227155809-63c882c7-d3bd-4039-b1cf-d7bc4f7a46e3.png)
16. Add Adjust contact points action using another Group
![image](https://user-images.githubusercontent.com/8580942/227156188-464eb6c1-32d9-4f1f-b75e-9704dc229bb0.png)
![image](https://user-images.githubusercontent.com/8580942/227156163-d0b6f474-9c5c-42a1-a5ff-075a00a13a35.png)
17. Run campaign update and trigger `ddev exec bin/console m:c:u && ddev exec bin/console m:c:t`
18. Check that contacts in the campaign have second Group points assigned
![image](https://user-images.githubusercontent.com/8580942/227156836-eb3dea68-e385-40f7-a075-ff84b8a98d6f.png)
19. Create new report with Group score as Datasource, add some columns (e.g. group name, group score, contact email) and check if report results are correct








<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->

patrykgruszka committed:
feat: [DPMMA-1132] score league entities

patrykgruszka committed:
feat: [DPMMA-1132] league migrations

patrykgruszka committed:
feat: [DPMMA-1132] league mvc

patrykgruszka committed:
feat: [DPMMA-1132] league permissions

patrykgruszka committed:
feat: [DPMMA-1132] add leagues to points and triggers

patrykgruszka committed:
feat: [DPMMA-1132] adjust point action with league

patrykgruszka committed:
feat: [DPMMA-1132] point triggers with league

patrykgruszka committed:
feat: [DPMMA-1132] point action functional tests

patrykgruszka committed:
feat: [DPMMA-1132] point trigger for existing contacts functional tests

patrykgruszka committed:
feat: [DPMMA-1132] lead point leagues view

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1132_point-leagues

npracht commented:
Awesome feature !
Ping me when ready to test, I'll do the end user tests :)

patrykgruszka committed:
feat: [DPMMA-1132] phpstan fix

patrykgruszka committed:
feat: [DPMMA-1132] FormExitSubscriber prop type fix

patrykgruszka committed:
fiz: [DPMMA-1132] unit tests fix

patrykgruszka committed:
feat: [DPMMA-1998] point league campaign decision

patrykgruszka committed:
feat: [DPMMA-1998] point league campaign decision comparison

patrykgruszka committed:
feat: [DPMMA-1998] league score repo functional tests

patrykgruszka committed:
feat: [DPMMA-1998] campaign subscriber functional tests

patrykgruszka committed:
feat: [DPMMA-1998] point league campaign decision messages

patrykgruszka committed:
fix: [DPMMA-1998] fix unit tests

RCheesley commented:
@patrykgruszka setting this to draft until you are ready for testers.

patrykgruszka committed:
Merge branch '5.x' into DPMMA-2002_point-leagues-main

patrykgruszka committed:
fix: [DPMMA-2002] lead twig template

patrykgruszka committed:
fix: [DPMMA-1132] add mautic.point.permissions.leagues message

patrykgruszka committed:
fix: [DPMMA-1132] update point bundle entities

patrykgruszka committed:
fix: [DPMMA-1132] add point league hint

patrykgruszka committed:
fix: [DPMMA-1132] update migration

patrykgruszka committed:
fix: [DPMMA-1999] migration down fix

patrykgruszka committed:
fix: [DPMMA-1998] lead model fix

patrykgruszka committed:
fix: [DPMMA-1999] lead point change campaign action

patrykgruszka committed:
fix: [DPMMA-1999] fix undefined array key

patrykgruszka committed:
fix: [DPMMA-1999] compare score select

patrykgruszka committed:
fix: [DPMMA-1999] points log event name

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1132_point-leagues

patrykgruszka committed:
fix: [DPMMA-1999] leagues twig templates in point bundle

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1132_point-leagues

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1132_point-leagues

patrykgruszka committed:
fix: [DPMMA-2000] league form template fix after merge

patrykgruszka committed:
fix: [DPMMA-2000] league form template fix

patrykgruszka committed:
fix: [DPMMA-2000] allow use of contact_id column in segment filters

patrykgruszka committed:
feat: [DPMMA-2000] league score segment filter

patrykgruszka committed:
fix: [DPMMA-2000] refactor foreign contact cols in segment filters

patrykgruszka committed:
feat: [DPMMA-2000] segment filter tests

patrykgruszka committed:
fix: [DPMMA-2000] fix tests with db prefix

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1132_point-leagues

patrykgruszka committed:
fix: [DPMMA-2001] remove unnecessary fields from league_contact_score table

patrykgruszka committed:
fix: [DPMMA-2001] add league fk to lead_point_change_log

patrykgruszka committed:
fix: [DPMMA-2000] get foreign contact column for all operators

patrykgruszka committed:
feat: [DPMMA-2001] add league column to contact point log report

patrykgruszka committed:
feat: [DPMMA-2001] lead score report

patrykgruszka committed:
fix: [DPMMA-2001] fixes and league report functional tests

patrykgruszka committed:
fix: [DPMMA-2001] fix failing tests

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1132_point-leagues

patrykgruszka committed:
fix: [DPMMA-2058] league points update shoud not change main lead points

patrykgruszka committed:
fix: [DPMMA-2058] fix points change log and point triggers

patrykgruszka commented:
@npracht, @RCheesley this feature is ready to test :)

npracht commented:
🚀 !

I'll give a try asap :)

escopecz commented:
I'm a bit confused from the naming. It may be because English is my second language but I wouldn't consider "league" being a synonym for "category". Although Thesaurus says it is. How about "group" or "class"?

@RCheesley what wording would you use?

RCheesley commented:
It's a difficult one - league is like what we have in sports for example - the premier league. Like an ordered list of where a team is in the rankings. I don't think that exactly fits what we are trying to do here.

I wonder if 'Point Types' would be a better term to use?

As a marketer I would imagine that you would use this feature to create different types of points for different kinds of things - for example sales-related actions, support-related actions - or maybe types of products that the person is engaging with.

What do you think @patrykgruszka ? Any thoughts @jos0405 @fakela ?

patrykgruszka commented:
I thought a lot about the name before implementing this functionality, and leagues seemed the most appropriate. Points are actually a kind of ranking, but I am open to other suggestions.

npracht commented:
Add emoji (reaction) to post as following to vote:
👍🏽 `Point Types`,
😁 `Point Families`,
🚀  `Point Groups`,
❤️  `Point Sections`,
🎉  `Point Tiers`
👀  `Point Leagues`

atowne-mautic commented:
Should the voting for the name on this be carried over to a forum/slack, where more of the marketing users will see it and be able to provide input? I agree with John that "Leagues" wouldn't be the best option. Categories, Groups, or Types all sound much better.

Perhaps it'd be a separate ticket, but would we also want to consider adding options for demographic scoring as a Point Action? For example, a user could set a Point Action to say "If a contact's position contains VP or Vice President, add 10 points to the Profile Points category/league". Currently that's possible through Campaigns, but would be much easier in point actions (and if I've missed another ticket about this, please share!).

patrykgruszka commented:
Honestly, I would like the Mautic team to decide which option is the best so I can make the adjustments if needed, and get it ready to test.

RCheesley commented:
I have requested @YosuCadilla to put this out for community review, as the majority of the votes to date appear to be from the business of the PR creator, therefore I feel we require wider engagement on this.

YosuCadilla commented:
Hi @RCheesley I'm ready to advance this consultation, but I am stuck on the comms hub

YosuCadilla commented:
Hello @RCheesley I'm ready to advance this consultation, but I am stuck on the comms hub, how should I proceed?

RCheesley commented:
> Should the voting for the name on this be carried over to a forum/slack, where more of the marketing users will see it and be able to provide input?

I've posted this to the forums and pinned it for the rest of the week. I basically added the comment onto the 4+ year old feature request for this feature, pointing folks to the the message with the emojis.

Let's make a decision one way or the other by Monday next week (19th June).

jos0405 commented:
Hi, just wanted to mention, that many marketers use points in the tracking gif.
Would it be possible to have points added based on page visit?
mtracking.gif?pointleageA=+10 ? or something similar?

patrykgruszka commented:
@jos0405 of course, this can be implemented once we have the basic functionality in place. This PR is already very complex.

npracht commented:
<img width="995" alt="Capture d’écran 2023-06-19 à 09 39 06" src="https://github.com/mautic/mautic/assets/14075239/1fd2f898-591e-489f-9e83-98a0c0fcd0cd">
I think we go for Point Groups 🚀 

@patrykgruszka could you please rename it in code + fix conflicts ? Then we can push it asap to release :)

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1132_point-leagues

patrykgruszka committed:
fix: [DPMMA-2242] csfixer and rector fixes

patrykgruszka committed:
fix: [DPMMA-2242] phpstan and test fixes

patrykgruszka committed:
fix: [DPMMA-2242] fix TriggerModelTest

RCheesley committed:
Merge branch '5.x' into DPMMA-1132_point-leagues

RCheesley commented:
Thanks for making the updates @patrykgruszka - let's get this tested and merged! 🚀

patrykgruszka commented:
Renaming is not yet done, I have just upgraded to 5.x and resolved conflicts.

kuzmany committed:
Fix template base for twig

kuzmany commented:
@patrykgruszka I fixed some templates issue after merge

kuzmany committed:
Fix old link to repo

kuzmany committed:
Change mauticContent naming

kuzmany committed:
Fix wrong language string

kuzmany commented:
@patrykgruszka it seems like this PR does not contain segment filter. I am forking it, but my segment filters for leagues cannot load. Can you point me to lines where you added it?

patrykgruszka commented:
@kuzmany sure, it is in this file `app/bundles/PointBundle/EventListener/SegmentFilterSubscriber.php:48`. You need to add a league first, because each league is a new filter due to the configuration limit (I mean, you can't select league and condition). I'm not sure if everything works fine after this latest update, as I had to make a lot of adjustments due to phpstan and templates. I will change the 'league' to 'group' and then test again to make sure everything works well.

patrykgruszka committed:
fix: [DPMMA-2242] rename league to group

patrykgruszka committed:
fix: [DPMMA-2242] csfixer changes

patrykgruszka commented:
Ok, I have changed the name of the feature and tested it to see if it works correctly. I've found some problems with the contact filters, but this is not related to this PR and exists in 5.x.

@kuzmany thanks for your help with the templates!

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1132_point-leagues

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1132_point-leagues

patrykgruszka commented:
The contact list filtering issue has been fixed. I have also updated the user documentation after changing the name to Point Groups.

@npracht this feature is ready to test.

RCheesley reviewed (changes requested):
Here's a small bug I found - missing language string:

![screenshot-8080-patrykgruszka-mautic-wy4h71n1tcc ws-eu100 gitpod io-2023 06 22-19_57_51](https://github.com/mautic/mautic/assets/2930593/7ed03a57-c262-4a47-850d-ebe7d44eaa10)

I found this a little confusing (only one points field) and not sure if the layout bug is relating to this PR or not:
![screenshot-8080-patrykgruszka-mautic-wy4h71n1tcc ws-eu100 gitpod io-2023 06 22-20_20_03](https://github.com/mautic/mautic/assets/2930593/9f8693ad-9018-42e3-90e4-bc2c11f48447)

Also here I think we need to explain if this is points of any group, or just the total points across all:

![screenshot-8080-patrykgruszka-mautic-wy4h71n1tcc ws-eu100 gitpod io-2023 06 22-20_22_02](https://github.com/mautic/mautic/assets/2930593/1bed4ed5-70f4-4056-a893-e29f18778dbd)

I created a trigger with a condition of points being above 40 and selected to have it apply to all existing contacts, set it on add/remove from a segment, but a contact with 40 points did not change:
![screenshot-8080-patrykgruszka-mautic-wy4h71n1tcc ws-eu100 gitpod io-2023 06 22-20_25_02](https://github.com/mautic/mautic/assets/2930593/1797736a-1abf-4c18-8f7f-8471330db165)

![screenshot-8080-patrykgruszka-mautic-wy4h71n1tcc ws-eu100 gitpod io-2023 06 22-20_29_06](https://github.com/mautic/mautic/assets/2930593/fd7b136e-e30f-46d5-81c6-95112d32ed5a)

This is what I found so far in testing, hope that's helpful!

kuzmany commented:
Yes. triggers not working for groups.  Other features looks good to me, I've tested it last days

patrykgruszka committed:
fix: [DPMMA-2242] fix point triggers for groups

patrykgruszka commented:
@kuzmany, @RCheesley thanks for your tests, I've fixed the problem with the triggers and it should work fine now. 

I've also added tooltip next to the group select, with text: "Choose the point group to which this applies. If no groups are selected, it will apply to the main contact points.".

@RCheesley In your case, the trigger required at least 40 points in group **Test D**. The contact had 50 main points, so it should not be triggered. 

At the moment it is not possible to manually set points for each group in the contact edit form. This is a nice feature in my opinion and could be added in the future.

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1132_point-leagues

patrykgruszka committed:
fix: [DPMMA-2242] point grop model alias

kuzmany committed:
Fix set group if null

patrykgruszka committed:
fix: [DPMMA-2242] csfixer

magdalenaleonow reviewed (changes requested):
Tested, found another language string missing:
![image](https://github.com/mautic/mautic/assets/130159751/13b8984f-142f-4b34-a56b-e9511a69ea21)
Everything else works great:
![image](https://github.com/mautic/mautic/assets/130159751/4effb0f3-ae36-48d4-898b-d275ada5d967)
![image](https://github.com/mautic/mautic/assets/130159751/2204bcae-8074-4840-922b-5ba856348985)

patrykgruszka committed:
fix: [DPMMA-2252] add missing translation

patrykgruszka commented:
@magdalenaleonow thanks for testing, I've added the missing translation.

magdalenaleonow reviewed (approved):
@patrykgruszka everything's ok now

kuzmany committed:
Add support for point groups for form action

kuzmany commented:
@patrykgruszka added support for point groups for form action

kuzmany committed:
Fix PHP Stan

escopecz commented on app/bundles/LeadBundle/Entity/Lead.php:
```diff
@@ -377,6 +385,12 @@ public static function loadMetadata(ORM\ClassMetadata $metadata)
             ->fetchExtraLazy()
             ->build();
 
+        $builder->createOneToMany('groupScores', 'Mautic\PointBundle\Entity\GroupContactScore')
```
It's better to use constants over strings.
```suggestion
        $builder->createOneToMany('groupScores', \Mautic\PointBundle\Entity\GroupContactScore::class)
```

escopecz replied:
Also, I don't see this column being added in the migration.

escopecz replied:
One more question, is there a reason for camelCase? All the other columns are using underscores as word separators.

patrykgruszka replied:
This is not the column name, it describes the one-to-many relationship with the `point_group_contact_score` table just like the `utmtags`. I can't see any use of snake_case when calling `createOneToMany` method.

Version20230621074925.php:64
```
$this->addSql("CREATE TABLE `{$contactScoreTableName}`
(
    `contact_id`          BIGINT UNSIGNED NOT NULL,
    `group_id`           INT UNSIGNED    NOT NULL,
    `score`               INT             NOT NULL,
    PRIMARY KEY (`contact_id`, `group_id`)

...

$this->addSql("ALTER TABLE `{$contactScoreTableName}` ADD CONSTRAINT `{$contactScoreContactFk}` FOREIGN KEY (`contact_id`) REFERENCES `{$contactTableName}` (`id`) ON DELETE CASCADE");

```

escopecz commented on app/bundles/LeadBundle/Entity/Lead.php:
```diff
@@ -878,7 +892,7 @@ public function resetPointChanges()
      * Creates a points change entry.
      */
-    public function addPointsChangeLogEntry($type, $name, $action, $pointChanges, IpAddress $ip)
+    public function addPointsChangeLogEntry($type, $name, $action, $pointChanges, IpAddress $ip, Group $group = null)
```
When changing a public method signature, which is considered a BC breaking change, please add the remaining param and return types.

patrykgruszka replied:
Even a change to a compatible signature with an optional parameter should be considered a BC? Why do you think so?

escopecz replied:
It is a BC breaking change either way. I'm suggesting that when we do this change, add all the missing types so we don't have to do it again later on.

escopecz commented on app/bundles/LeadBundle/Entity/Lead.php:
```diff
@@ -1975,4 +1992,32 @@ function ($a, $b) {
+    /**
+     * @return ArrayCollection<int,GroupContactScore>
+     */
+    public function getGroupScores()
```
```suggestion
    public function getGroupScores(): Collection
```

escopecz commented on app/bundles/LeadBundle/Form/Type/CampaignEventPointType.php:
```diff
@@ -0,0 +1,81 @@
+    public function __construct(TypeOperatorProviderInterface $typeOperatorProvider)
+    {
+        $this->typeOperatorProvider = $typeOperatorProvider;
+    }
```
Feel free to use shortened controllers in M5
```suggestion
    public function __construct(private TypeOperatorProviderInterface $typeOperatorProvider)
    {
    }
```

escopecz commented on app/bundles/LeadBundle/Form/Type/CampaignEventPointType.php:
```diff
@@ -0,0 +1,81 @@
+    public function getBlockPrefix()
+    {
+        return 'campaignevent_point';
+    }
```
This can be removed. It's present only in old form types for legacy purposes.

escopecz commented on app/bundles/LeadBundle/Segment/ContactSegmentFilter.php:
```diff
@@ -117,6 +117,18 @@ public function getTable()
+    /**
+     * @return string|null
+     */
+    public function getForeignContactColumn()
```
```suggestion
    public function getForeignContactColumn(): ?string
```

escopecz commented on app/bundles/LeadBundle/Segment/ContactSegmentFilter.php:
```diff
@@ -117,6 +117,18 @@ public function getTable()
+     */
+    public function getForeignContactColumn()
+    {
+        if (method_exists($this->filterDecorator, 'getForeignContactColumn')) {
```
Perhaps an interface would be a better way to go instead of checking if a method exists?

escopecz commented on app/bundles/LeadBundle/Segment/Decorator/BaseDecorator.php:
```diff
@@ -149,4 +149,14 @@ public function getWhere(ContactSegmentFilterCrate $contactSegmentFilterCrate)
+     *
+     * @return string
+     */
+    public function getForeignContactColumn(ContactSegmentFilterCrate $contactSegmentFilterCrate)
```
```suggestion
     */
    public function getForeignContactColumn(ContactSegmentFilterCrate $contactSegmentFilterCrate): string
```

escopecz commented on app/bundles/LeadBundle/Segment/Decorator/CustomMappedDecorator.php:
```diff
@@ -94,4 +94,20 @@ public function getWhere(ContactSegmentFilterCrate $contactSegmentFilterCrate)
+     *
+     * @return string
+     */
+    public function getForeignContactColumn(ContactSegmentFilterCrate $contactSegmentFilterCrate)
```
```suggestion
     */
    public function getForeignContactColumn(ContactSegmentFilterCrate $contactSegmentFilterCrate): string
```

escopecz commented on app/bundles/PointBundle/Entity/Group.php:
```diff
@@ -0,0 +1,99 @@
+    /**
+     * @var string|null
+     */
+    private $description;
```
```suggestion
    private ?int $id = null;

    private ?string $name;

    private ?string $description;
```

escopecz commented on app/bundles/PointBundle/Entity/Group.php:
```diff
@@ -0,0 +1,99 @@
+        $builder = new ClassMetadataBuilder($metadata);
+
+        $builder->setTable(self::TABLE_NAME)
+            ->setCustomRepositoryClass('Mautic\PointBundle\Entity\GroupRepository');
```
```suggestion
            ->setCustomRepositoryClass(\Mautic\PointBundle\Entity\GroupRepository::class);
```

escopecz commented on app/bundles/PointBundle/Entity/Group.php:
```diff
@@ -0,0 +1,99 @@
+    /**
+     * @return int
+     */
+    public function getId()
```
```suggestion
    public function getId(): ?int
```

escopecz commented on app/bundles/PointBundle/Entity/Group.php:
```diff
@@ -0,0 +1,99 @@
+     *
+     * @return self
+     */
+    public function setDescription($description)
```
```suggestion
    public function setDescription(string $description): self
```

escopecz commented on app/bundles/PointBundle/Entity/Group.php:
```diff
@@ -0,0 +1,99 @@
+    /**
+     * @return string
+     */
+    public function getDescription()
```
```suggestion
    public function getDescription(): ?string
```

escopecz commented on app/bundles/PointBundle/Entity/Group.php:
```diff
@@ -0,0 +1,99 @@
+     *
+     * @return self
+     */
+    public function setName($name)
```
```suggestion
    public function setName(string $name): self
```

escopecz commented on app/bundles/PointBundle/Entity/Group.php:
```diff
@@ -0,0 +1,99 @@
+    /**
+     * @return string
+     */
+    public function getName()
```
```suggestion
    public function getName(): ?string
```

escopecz commented on app/bundles/PointBundle/Entity/GroupContactScore.php:
```diff
@@ -0,0 +1,70 @@
+        $builder = new ClassMetadataBuilder($metadata);
+
+        $builder->setTable(self::TABLE_NAME)
+            ->setCustomRepositoryClass('Mautic\PointBundle\Entity\GroupContactScoreRepository');
```
```suggestion
            ->setCustomRepositoryClass(\Mautic\PointBundle\Entity\GroupContactScoreRepository:class);
```

escopecz commented on app/bundles/PointBundle/Entity/GroupContactScore.php:
```diff
@@ -0,0 +1,70 @@
+
+        $builder->addContact(false, 'CASCADE', true, 'group_score');
+
+        $builder->createManyToOne('group', 'Mautic\PointBundle\Entity\Group')
```
```suggestion
        $builder->createManyToOne('group', \Mautic\PointBundle\Entity\Group::class)
```

escopecz commented on app/bundles/PointBundle/Entity/GroupContactScore.php:
```diff
@@ -0,0 +1,70 @@
+            ->addJoinColumn('group_id', 'id', true, false, 'CASCADE')
+            ->build();
+
+        $builder->createField('score', 'integer')
```
```suggestion
        $builder->createField('score', Types::INTEGER)
```

escopecz commented on app/bundles/PointBundle/Entity/GroupContactScore.php:
```diff
@@ -0,0 +1,70 @@
+            ->build();
+    }
+
+    public function getContact(): Lead
```
All getters for Doctrine entities must return a nullable value unless a non-nullable value is set as default property value or required via constructor. I suggest to set required properties via constructor instead of setters so it is obvious.

escopecz commented on app/bundles/PointBundle/Entity/Point.php:
```diff
@@ -379,4 +385,14 @@ public function getRepeatable()
+        return $this->group;
+    }
+
+    public function setGroup(?Group $group = null): void
```
```suggestion
    public function setGroup(?Group $group): void
```
I can't imagine why would anyone call a setter with no value.

kuzmany committed:
Fix foreign key column

kuzmany commented:
@patrykgruszka your changes for segment filter stop work for email sent date filters. I fixed it with https://github.com/mautic/mautic/pull/11891/commits/be1c9614c2fd98e15ec496d358c44d23d3bd80ce. Can you review it?

patrykgruszka commented:
Ah, right. Thanks for the fix, looks good 👍. Using `contact_id` instead of `lead_id` (which is deprecated) caused a lot of unexpected work, but hopefully all issues are resolved now.

Failing Rector doesn't seem to be related to this pull request.

escopecz commented:
Failing Rector is being fixed in https://github.com/mautic/mautic/pull/12556

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1132_point-leagues

kuzmany reviewed (dismissed):
I spent some time testing it.
Seems good to me

patrykgruszka committed:
fix: [DPMMA-2252] code refactor after CR

patrykgruszka commented:
@escopecz thank you for your suggestions. They have been included in the latest changes. 

@magdalenaleonow could you please test this feature again?

escopecz commented on app/bundles/PointBundle/Entity/TriggerEventRepository.php:
```diff
@@ -31,6 +32,45 @@ public function getPublishedByPointTotal($points)
+     *
+     * @return mixed[]
+     */
+    public function getPublishedByGroupScore($groupScores)
```
```suggestion
    public function getPublishedByGroupScore(Collection $groupScores)
```

escopecz commented on app/bundles/PointBundle/Event/GroupEvent.php:
```diff
@@ -0,0 +1,38 @@
+use Mautic\CoreBundle\Event\CommonEvent;
+use Mautic\PointBundle\Entity\Group;
+
+class GroupEvent extends CommonEvent
```
Is it necessary to extend the `CommonEvent`? I'd like to get rid of any unnecessary inheritance.
```suggestion
final class GroupEvent
```

patrykgruszka replied:
Probably not, I've used this class to be consistent. Let's remove it if it's not necessary.

escopecz commented on app/bundles/PointBundle/Event/GroupEvent.php:
```diff
@@ -0,0 +1,38 @@
+
+class GroupEvent extends CommonEvent
+{
+    /**
+     * @param bool $isNew
+     */
+    public function __construct(Group &$group, $isNew = false)
+    {
+        $this->entity = &$group;
+        $this->isNew  = $isNew;
```
`$isNew` doesn't seem to be used. Is it necessary to pass `$group` as pointer? It is already a pointer for the object type, right?
```suggestion
    public function __construct(private Group $group, $isNew = false)
    {
```

patrykgruszka replied:
Right, I don't think that's necessary. It looked like there was some kind of pattern to creating these kinds of events in Mautic.

escopecz commented on app/bundles/PointBundle/Event/GroupEvent.php:
```diff
@@ -0,0 +1,38 @@
+        $this->isNew  = $isNew;
+    }
+
+    /**
+     * Returns the Group entity.
+     *
+     * @return Group
+     */
+    public function getGroup()
```
```suggestion
    public function getGroup(): Group
```

escopecz commented on app/bundles/PointBundle/Event/GroupScoreChangeEvent.php:
```diff
@@ -0,0 +1,53 @@
+use Mautic\LeadBundle\Entity\Lead;
+use Mautic\PointBundle\Entity\GroupContactScore;
+
+class GroupScoreChangeEvent extends CommonEvent
```
```suggestion
final class GroupScoreChangeEvent
```

escopecz commented on app/bundles/PointBundle/Event/GroupScoreChangeEvent.php:
```diff
@@ -0,0 +1,53 @@
+
+class GroupScoreChangeEvent extends CommonEvent
+{
+    protected int $oldScore;
+    protected int $newScore;
+
+    public function __construct(GroupContactScore &$groupContactScore, int $oldScore, int $newScore)
+    {
+        $this->entity      = &$groupContactScore;
+        $this->oldScore    = $oldScore;
+        $this->newScore    = $newScore;
+    }
```
```suggestion
    public function __construct(private GroupContactScore $groupContactScore, private int $oldScore, private int $newScore)
    {
    }
```

escopecz commented on app/bundles/PointBundle/Event/GroupScoreChangeEvent.php:
```diff
@@ -0,0 +1,53 @@
+        $this->newScore    = $newScore;
+    }
+
+    /**
+     * Returns the GroupContactScore enitity.
+     *
+     * @return GroupContactScore
+     */
+    public function getGroupContactScore()
```
```suggestion
    public function getGroupContactScore(): GroupContactScore
```

escopecz commented on app/bundles/PointBundle/Form/Type/GroupListType.php:
```diff
@@ -0,0 +1,79 @@
+    {
+        $this->em         = $em;
+        $this->repo       = $repo;
+    }
```
```suggestion
    public function __construct(private EntityManager $em, private GroupRepository $repo)
    {
    }
```

escopecz commented on app/bundles/PointBundle/Form/Type/GroupListType.php:
```diff
@@ -0,0 +1,79 @@
+    public function getBlockPrefix()
+    {
+        return 'group';
+    }
```
unnecessary
```suggestion
```

escopecz commented on app/bundles/PointBundle/Form/Type/GroupListType.php:
```diff
@@ -0,0 +1,79 @@
+    /**
+     * @return string
+     */
+    public function getParent()
```
```suggestion
    public function getParent(): string
```

escopecz commented on app/bundles/PointBundle/Resources/views/Trigger/_list.html.twig:
```diff
@@ -89,6 +96,11 @@
                         </span>
                     </td>
+                    <td class="visible-md visible-lg">
+                        {% set group = item.group %}
+                        {% set groupName = group.name|default('mautic.point.group.form.nogroup'|trans) %}
+                        {{ groupName }}
```
Do the variables have some deeper meaning?
```suggestion
                        {{ item.group.name|default('mautic.point.group.form.nogroup'|trans) }}
```

patrykgruszka replied:
I don't think so, let's clean it up

escopecz commented on app/bundles/PointBundle/Tests/Unit/Model/TriggerModelTest.php:
```diff
@@ -124,35 +122,9 @@ public function testTriggerEvent(): void
             ->method('dispatch')
             ->withConsecutive(
-                [
-                    $this->callback(
-                        // Emulate a subscriber:
-                        function (TriggerBuilderEvent $event) {
```
I know this was this complicated due to Swiftmailer. It should be easier with Symfony Mailer. But does it have to be removed without a replacement?

patrykgruszka replied:
You are right, I forgot to fix this test properly

escopecz commented on app/migrations/Version20230621074925.php:
```diff
@@ -0,0 +1,121 @@
+     */
+    public function preUp(Schema $schema): void
+    {
+        $shouldRunMigration = !$schema->hasTable($this->generateTableName(Group::TABLE_NAME));
+
+        if (!$shouldRunMigration) {
+            throw new SkipMigration('Schema includes this migration');
+        }
```
We have bad experience with these massive migrations and simple preUp checks. It fails in the middle for 1 instance. So our dev ops team run it again. It succeeds because the preUp check says it's migrated. But it's missing some alter tables that has failed before. So this migration silently fails and then our customers have to report broken instances.

That's why we created https://github.com/mautic/mautic/blob/9a609d5b8d83ddfa27bb752a91b863738f60b12c/app/migrations/Migration.template#L12 so it at least fails and we know about it.

patrykgruszka replied:
Good to know 👍 I will update this migration

escopecz reviewed (changes requested):
It seems I got distracted in the last review and did not review the second half. Apologies .

patrykgruszka committed:
fix: [DPMMA-2252] code refactor - migration and events

patrykgruszka committed:
fix: [DPMMA-2252] fix trigger model test

patrykgruszka commented:
@escopecz I've added your suggestions. Let me know if we are ready to test this again.

escopecz committed:
Fixing "$contactScoreTableName must not be accessed before initialization"

for
bin/console doctrine:migrations:execute "Mautic\Migrations\Version20230621074925" --down

escopecz committed:
Fixing setting null to a string param when saving a group form with no name or description

escopecz commented:
I'm wondering how will this work for webhooks? Still the same?
![Screenshot 2023-07-18 at 16 07 06](https://github.com/mautic/mautic/assets/1235442/bf3103f4-7e81-4d5a-a78f-42eeebf6e820)

escopecz commented:
The campaign action seem to work for original points but not for the group points. Here is the campaign action:
![Screenshot 2023-07-18 at 16 17 58](https://github.com/mautic/mautic/assets/1235442/c53eee51-80bf-4cd2-a7eb-4d987066da09)
You can see it should add 2 points to the contact but it still has 0 points even though it passed through the campaign action:
![Screenshot 2023-07-18 at 16 18 33](https://github.com/mautic/mautic/assets/1235442/88180745-a37a-44bd-87ec-2cc4853e1e77)
It would be really great to cover this with a functional test. Same with saving the group. There is no test that would discover that saving it with no description fails.

kuzmany commented:
> I'm wondering how will this work for webhooks? Still the same? ![Screenshot 2023-07-18 at 16 07 06](https://user-images.githubusercontent.com/1235442/254289924-bf3103f4-7e81-4d5a-a78f-42eeebf6e820.png)

It will not work for webhooks. Point groups do not come to contact entity changes and then cannot work. 
This is the reason why also point triggers do not work for groups.

patrykgruszka commented:
@escopecz I've tested the points action with group in campaign with fresh install on Gitpod and it works fine for me. Seems like I need more details to replicate your problem.
![image](https://github.com/mautic/mautic/assets/8580942/864d17bc-67b5-4dd2-8fdd-12e0569e479b)

@kuzmany I've fixed the problem with Group Point Triggers some time ago in https://github.com/mautic/mautic/pull/11891/commits/6598baf6e5b7b19591069697c431d33784b47bb5. I'm not sure about the webhooks, need to check that.

kuzmany commented:
@patrykgruszka your commit does not fix the trigger on point change. If points are changed for groups for example with form, trigger not execute. It cannot be because it runs based on changes of contacts and these points not coming there

patrykgruszka commented:
> I'm wondering how will this work for webhooks? Still the same? ![Screenshot 2023-07-18 at 16 07 06](https://user-images.githubusercontent.com/1235442/254289924-bf3103f4-7e81-4d5a-a78f-42eeebf6e820.png)

The webhook will be triggered only on contact points change, when `LeadEvents::LEAD_POINTS_CHANGE` is triggered. We need to define a new webhook or modify the existing one to listen for `PointGroupEvents::SCORE_CHANGE` event. Do you think that this is necessary to implement in this PR or could be added later as a new feature?

escopecz commented:
> Do you think that this is necessary to implement in this PR or could be added later as a new feature?

It doesn't have to be implemented in this PR. It can be done in a future PR. It's not breaking the original point implementation so it will work for whoever is using it already. But please drop a note to the docs PR for this feature about the webhooks and that the group points won't trigger the webhook.

escopecz commented:
> Seems like I need more details to replicate your problem.

@patrykgruszka I think I see the issue. Here are steps:
1. Create a new campaign
2. Add the Point action but do not set the group
3. Save&close
4. Edit the point action and add the group
5. You can see in the campaign action that the group is empty in the `properties` array but set in the root of the campaign action:
```php
Array
(
    [id] => new159
    [name] => 
    [description] => 
    [type] => lead.changepoints
    [eventType] => action
    [order] => 2
    [properties] => Array
        (
            [points] => -2
            [group] => 
        )

...

    [points] => -2
    [group] => 1
)
```
The subscriber takes the group from the `properties` array: https://github.com/mautic/mautic/pull/11891/files#diff-40d122381bd5c9c8d3746934611aacdbe9af898420d3a2f8cb5e613dad7ea806R323

patrykgruszka committed:
fix: [DPMMA-2252] fix group points persistence in events

patrykgruszka commented:
@escopecz, Yes, that's weird, but I've tested it with the debugger and it does assign the correct value. The nested properties are ignored here. Most of the campaign actions use `$event->getConfig()` and the values returned are correct, although there is an incorrect value in `$event['properties']['properties']`. This problem doesn't seem to be related to this PR.
![image](https://github.com/mautic/mautic/assets/8580942/c97e479b-32a8-42b7-8684-39941c1a9601)
Correct values: `$points=3` `$pointGroupId=3`, outdated properties are ignored.

In your case, Group A is set to 0 points, so something must have happened to the group points. Maybe a negative value was before that action? Unfortunately, I still can't reproduce this problem.

@kuzmany I found a persistence issue that caused the points to not be assigned correctly. I've tested form action triggers with point group and it works fine now.

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1132_point-leagues

patrykgruszka commented:
> > Do you think that this is necessary to implement in this PR or could be added later as a new feature?
> 
> It doesn't have to be implemented in this PR. It can be done in a future PR. It's not breaking the original point implementation so it will work for whoever is using it already. But please drop a note to the docs PR for this feature about the webhooks and that the group points won't trigger the webhook.

I've added this note to the docs and info about the Form action with group points implemented by @kuzmany.

escopecz reviewed (approved):
I must have had an older version on my local. It started working when I pulled 👍 

Thanks for making this great feature!

LeonOltmanns reviewed (changes requested):
Just tested this amazing PR and found two little things, which I believe @RCheesley also already mentioned:

<img width="1417" alt="PR-PG-company+pointsfields" src="https://github.com/mautic/mautic/assets/55587275/3998b952-8bb2-4680-aa14-15987622a7f6">

1. There only is the option to manually add points to the "default/main" score, but not for the point groups
2. The primary company field is, for whatever reason, full width. I used a fresh mautic instance and solely tested this PR, so I dont think any other PR should make this change besides the Point Groups one

Besides this two minor findings its an awesome feature, thank you so much!

magdalenaleonow reviewed (approved):
@patrykgruszka tested, works great!

jos0405 reviewed (approved):
Amazing feature, so much needed.
Works well.

Tested URL visit based point gaining:
![image](https://github.com/mautic/mautic/assets/4246909/38998534-2521-464c-afad-7192782fe384)

Also Email based point gains:
![image](https://github.com/mautic/mautic/assets/4246909/2b8a0d45-b1de-4c78-a9da-b9597d57db52)

Segment, campaign condition, action works also well:
![image](https://github.com/mautic/mautic/assets/4246909/fb5027c6-c30c-4255-a21a-bb4f5107b57f)

jos0405 commented:
@patrykgruszka , this is an amazing feature, it should be added to Mautic 5 for sure as soon as possible.
The point about editing the points from the contact profile is realy valid, can you make it work?
https://github.com/mautic/mautic/pull/11891#pullrequestreview-1541203734
During the Sprint in Wien we decided to add this feature to Beta, it would be a clear loss no to, but it would be wonderful if you could add that missing piece for the RC.

escopecz committed:
Merge branch '5.x' into DPMMA-1132_point-leagues
